### PR TITLE
Improvements to libcloud modules

### PR DESCRIPTION
**Pull request by tonybaloney**
# Improvements to libcloud modules

## libcloud_dns module

- Add filtering for DNS record by type
- Return simple dictionaries so the user doesn't need to use a custom renderer
- Add support for extended methods

## libcloud_storage and libcloud_loadbalancer

- Added extra method, add support for kwargs being passed to the libcloud methods

## docs

- Added tutorial for the libcloud modules
- Added doc module for the state modules

## todo

Complete state examples and allow custom arguments to be passed from states. This will come in a later PR, this is usable already

### CODE REVIEW

**Comment by tonybaloney** [comment]
demonstrating functionality of the end-to-end state modules
```
salt-call -c etc/salt state.apply --local
[ERROR   ] Created new container
local:
    ----------
    libcloud_storage_|-storage_test_|-example_salty_|-container_present:
        ----------
        __id__:
            storage_test
        __run_num__:
            0
        __sls__:
            storage_test
        changes:
            ----------
        comment:
            Created new container
        duration:
            2300.623
        name:
            example_salty
        result:
            ----------
            extra:
                ----------
            name:
                example_salty
        start_time:
            16:39:11.365541
(salt) (python2) Anthonys-MacBook-Pro-2:salt anthonyshaw$ cat etc/salt/states/storage_test.sls 
storage_test:
    libcloud_storage.container_present:
        - name: example_salty
        - profile: google
```

**Comment by tonybaloney** [comment]
```
local:
----------
          ID: storage_test
    Function: libcloud_storage.container_present
        Name: example_salty2
      Result: True
     Comment: Created new container
     Started: 16:42:37.840175
    Duration: 2889.835 ms
     Changes:   
              ----------
              extra:
                  ----------
              name:
                  example_salty2

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:   2.890 s
```

**Comment by tonybaloney** [comment]
also @techhat for review

**Comment by tonybaloney** [comment]
That last failure is unrelated @cachedout @rallytime this one is ready to merge when reviewed

**Comment by tonybaloney** [comment]
demo using the `libcloud_loadbalancer.balancer_present` state with GCE and storage containers

```
local:
----------
          ID: storage_test
    Function: libcloud_storage.container_present
        Name: example_salty2
      Result: True
     Comment: Container already exists
     Started: 20:55:00.727931
    Duration: 672.1 ms
     Changes:   
----------
          ID: lb_test
    Function: libcloud_loadbalancer.balancer_present
        Name: example
      Result: True
     Comment: Created new load balancer
     Started: 20:55:01.402015
    Duration: 13810.998 ms
     Changes:   
              ----------
              extra:
                  ----------
                  forwarding_rule:
                  healthchecks:
                  targetpool:
              id:
                  3511488495240462945
              ip:
                  104.196.126.238
              name:
                  example
              port:
                  80-80
              state:
                  None

Summary for local
------------
Succeeded: 2 (changed=1)
Failed:    0
------------
Total states run:     2
Total run time:  14.483 s
```

File: salt/modules/libcloud_loadbalancer.py [hunk 1]
```diff
@@ -380,3 +452,17 @@ def _simple_member(member):
         'balancer': _simple_balancer(member.balancer),
         'extra': member.extra
     }
+
+
+def _sanitize_kwargs(kwargs):
```
Comment (rallytime) [inline salt/modules/libcloud_loadbalancer.py]: We should already have a function for this in `salt.utils` called `clean_kwargs`. Maybe using that one over implementing a new version would be good here.

**Review by rallytime — COMMENTED** [review]
Very nice! I only had one suggestion for this PR.

File: doc/topics/tutorials/libcloud.rst [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+==============================================================================
+
+.. versionadded:: Oxygen
+
+.. note::
+
+    This walkthrough assumes basic knowledge of Salt and Salt States. To get up to speed, check out the
+    :ref:`Salt Walkthrough <tutorial-salt-walk-through>`.
+
+Apache Libcloud is a Python library which hides differences between different
```
Comment (cachedout) [inline doc/topics/tutorials/libcloud.rst]: Weird formatting here. Might be good to just make a hard stop at 80 chars for this whole section of test.

File: doc/topics/tutorials/libcloud.rst [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+            key: AFDDJFGIjdfgdfgdf
+            secret: FG(#f8vdfgjlkm)
+
+Using the execution modules
+===========================
+
+Amongst over 60 clouds that Apache Libcloud supports, you can add profiles to your Salt configuration to access and control these clouds. Each of the libcloud execution 
+modules exposes the common API methods for controlling DNS, Load Balancers and Object Storage. To see which functions are supported across specific clouds, see the Libcloud `supported methods <http://libcloud.readthedocs.io/en/latest/supported_providers.html#supported-methods-block-storage>`_ documentation.
+
+The module documentation explains each of the API methods and how to leverage them
```
Comment (cachedout) [inline doc/topics/tutorials/libcloud.rst]: Period at end of sentence.

File: doc/topics/tutorials/libcloud.rst [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+.. code-block:: bash
+
+    $ salt myminion libcloud_storage.create_balancer my_balancer 80 http profile1 ex_region=us-east1
+    $ salt myminion libcloud_storage.list_container_objects my_bucket profile1 ex_prefix=me
+
+Accessing custom methods in Libcloud drivers
+============================================
+
+Some cloud APIs have additional methods that are prefixed with `ex_` in Apache Libcloud, these methods are part of the non-standard API but can still
```
Comment (cachedout) [inline doc/topics/tutorials/libcloud.rst]: Same weird multi-line formatting problem here.

File: salt/modules/libcloud_loadbalancer.py (hunk 1 above)
Comment (tonybaloney) [inline salt/modules/libcloud_loadbalancer.py]: dang, I thought there might be! I just couldn't find it.

Comment (rallytime) [inline salt/modules/libcloud_loadbalancer.py]: Yeah, I think it's a relatively recent function. Plus it's buried in the massive `salt.utils.__init__.py` file. :)

Comment (tonybaloney) [inline salt/modules/libcloud_loadbalancer.py]: fixed.

File: doc/topics/tutorials/libcloud.rst (hunk 4 above)
Comment (tonybaloney) [inline doc/topics/tutorials/libcloud.rst]: fixed in e2dc7bf

**Comment by tonybaloney** [comment]
oops. the util one worked differently to mine. fixed now.

**Comment by tonybaloney** [comment]
@cachedout @rallytime ok I am *done* :-)

**Comment by tonybaloney** [comment]
@techhat you good with this one? We need to discuss compute. I left it out on purpose

**Review by techhat — APPROVED** [review]
Sorry, I hadn't realized I hadn't reviewed this one yet. LGTM!

**Comment by tonybaloney** [comment]
@cachedout @techhat @rallytime this one is done. I just had to resolve a new merge conflict.

**Comment by tonybaloney** [comment]
@s0undt3ch because of https://github.com/saltstack/salt/pull/40893/commits/df93290f24111bb6fa59d3f40fafdc16889c3f8a the actual module now responds with dictionaries instead of the libcloud classes so that it renders correctly in the salt CLI. I should probably just replace all the classes with dictionaries, but this was easier and I knew the tests work.

**Comment by tonybaloney** [comment]
fixed it properly..

**Comment by s0undt3ch** [comment]
Oh! Makes sense now @tonybaloney.

**Pull request merged**